File: src/main/java/org/library/LibraryDate.java
package org.library;

import java.util.Calendar;
import java.util.GregorianCalendar;

public class LibraryDate {
    private Calendar date;

    public LibraryDate(int year, int month, int day){
        date = new GregorianCalendar(year, month, day);
    }

    public LibraryDate(LibraryDate libraryDate){
        date = (Calendar)libraryDate.getDate().clone();
    }

    public LibraryDate(Calendar calendar){
        date = calendar;
    }

    public void addDays(int days){
        date.add(Calendar.DATE, days);
    }

    public Calendar getDate(){
        return date;
    }

    @Override
    public String toString() {
        int year = date.get(Calendar.YEAR);
        int month = date.get(Calendar.MONTH)+1;
        int day = date.get(Calendar.DAY_OF_MONTH);

        String monthStr = (month <= 9) ? "0"+month : ""+month;
        String dayStr = (day <= 9) ? "0"+day : ""+day;

        return String.format("%d-%s-%s", year, monthStr, dayStr);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (!(obj instanceof LibraryDate)) return false;
        LibraryDate libraryDate = (LibraryDate) obj;
        Calendar calendar = libraryDate.getDate();

        boolean equalYear = calendar.get(Calendar.YEAR) == date.get(Calendar.YEAR);
        boolean equalMonth = calendar.get(Calendar.MONTH) == date.get(Calendar.MONTH);
        boolean equalDay = calendar.get(Calendar.DAY_OF_MONTH) == date.get(Calendar.DAY_OF_MONTH);

        return equalYear && equalMonth && equalDay;
    }
}

File: main/LibraryDate.js
export default class LibraryDate {
  constructor(...args) {
    // Possible signatures:
    // new LibraryDate(year, month, day)
    // new LibraryDate(libraryDate)
    // new LibraryDate(date) where date is a JS Date
    if (args.length === 3 && typeof args[0] === "number") {
      const [year, month, day] = args;
      this.date = new Date(year, month, day);
    } else if (args.length === 1) {
      const arg = args[0];
      if (arg instanceof LibraryDate) {
        this.date = new Date(arg.getDate().getTime());
      } else if (arg instanceof Date) {
        this.date = new Date(arg.getTime());
      } else {
        // try to construct from an object with year/month/day
        const { year, month, day } = arg || {};
        if (
          typeof year === "number" &&
          typeof month === "number" &&
          typeof day === "number"
        ) {
          this.date = new Date(year, month, day);
        } else {
          this.date = new Date();
        }
      }
    } else {
      // default to current date
      this.date = new Date();
    }
  }

  addDays(days) {
    this.date.setDate(this.date.getDate() + days);
  }

  getDate() {
    return new Date(this.date.getTime());
  }

  toString() {
    const year = this.date.getFullYear();
    const month = this.date.getMonth() + 1; // JS months are 0-based
    const day = this.date.getDate();

    const monthStr = month <= 9 ? `0${month}` : `${month}`;
    const dayStr = day <= 9 ? `0${day}` : `${day}`;

    return `${year}-${monthStr}-${dayStr}`;
  }

  equals(obj) {
    if (obj === this) return true;
    if (!(obj instanceof LibraryDate)) return false;
    const calendar = obj.getDate();

    const equalYear = calendar.getFullYear() === this.date.getFullYear();
    const equalMonth = calendar.getMonth() === this.date.getMonth();
    const equalDay = calendar.getDate() === this.date.getDate();

    return equalYear && equalMonth && equalDay;
  }
}
